tensorboard/webapp/widgets/histogram/histogram_v2_component.ts

```diff
@@ -0,0 +1,193 @@
+import {
+  ChangeDetectionStrategy,
+  Component,
+  ElementRef,
+  Input,
+  OnChanges,
+  ViewChild,
+} from '@angular/core';
+
+import * as d3 from '../../third_party/d3';
+import {
+  ColorScale,
+  HistogramData,
+  HistogramMode,
+  TimeProperty,
+} from './histogram_types';
+
+@Component({
+  selector: 'tb-histogram-v2',
+  templateUrl: 'histogram_v2_component.ng.html',
+  styleUrls: ['histogram_v2_component.css'],
+  changeDetection: ChangeDetectionStrategy.OnPush,
+})
+export class HistogramV2Component implements OnChanges {
+  @ViewChild('xAxis') private readonly xAxis!: ElementRef;
+  @ViewChild('yAxis') private readonly yAxis!: ElementRef;
+  @ViewChild('content') private readonly content!: ElementRef;
+
+  @Input() mode: HistogramMode = HistogramMode.OFFSET;
+
+  @Input() timeProperty: TimeProperty = TimeProperty.STEP;
+
+  @Input() colorScale?: ColorScale;
+
+  @Input() name!: string;
+
+  @Input() data!: HistogramData;
+
+  private yScale?:
+    | d3.ScaleTime<number, number>
+    | d3.ScaleLinear<number, number>;
+
+  constructor(private readonly host: ElementRef) {}
+
+  ngOnChanges() {
+    if (this.content) {
+      this.ngAfterViewInit();
+    }
+  }
+
+  ngAfterViewInit() {
+    this.updateChart();
+  }
+
+  private updateChart() {
+    this.renderXAxis(this.data);
+    this.updateYScaleAndDomain(this.data);
+    this.renderYAxis();
+  }
+
+  private renderXAxis(data: HistogramData) {
+    const {width} = this.host.nativeElement.getBoundingClientRect();
+    const {min: xMin, max: xMax} = getMinMax(
+      data,
+      (datum) => getMin(datum.bins, (binVal) => binVal.x),
+      (datum) => getMax(datum.bins, ({x, dx}) => x + dx)
+    );
+    const xScale = d3
+      .scaleLinear()
+      .domain([xMin, xMax])
+      .nice()
+      .range([0, width]);
+    const xAxis = d3.axisBottom(xScale).ticks(Math.max(2, width / 20));
+
+    xAxis(d3.select(this.xAxis.nativeElement));
+  }
+
+  private updateYScaleAndDomain(data: HistogramData) {
+    this.yScale =
+      this.mode !== HistogramMode.OVERLAY &&
+      this.timeProperty == TimeProperty.WALL_TIME
+        ? d3.scaleTime()
+        : d3.scaleLinear();
+
+    let domain: [number, number] = [0, 1];
+    if (this.mode === HistogramMode.OVERLAY) {
+      const countMax = getMax(data, (datum) => {
+        return getMax(datum.bins, ({y}) => y);
+      });
+      domain = [0, countMax];
+    } else {
+      const yValues = data.map((datum) => {
+        switch (this.timeProperty) {
+          case TimeProperty.WALL_TIME:
+            return datum.wallTime;
+          case TimeProperty.STEP:
+            return datum.step;
+          case TimeProperty.RELATIVE:
+            return datum.wallTime - data[0]?.wallTime ?? 0;
+        }
+      });
+      const {min: yMin, max: yMax} = getMinMax(yValues, (val) => val);
+      domain = [yMin, yMax];
+    }
+    this.yScale.domain(domain);
+  }
+
+  private renderYAxis() {
+    const yScale = this.yScale!;
```

Comment:
I'm not sure whether followups need the `yScale` to be a property, but would it make more sense to make `yScale` a parameter to `renderYAxis` instead of asserting truthiness here?

Reply:
Will address that in a follow up that PR is about ready.